KerbalStuff/app.py

```diff
@@ -70,49 +69,77 @@ def load_user(username: str) -> User:
-            app.logger.addHandler(mail_handler)
+# Send ERROR and EXCEPTION level log entries to syslog and per email.
+error_to = _cfg("error-to")
+if error_to:
+    import logging
+    from logging.handlers import SMTPHandler
+
+    smtp_host = _cfg('smtp-host')
+    smtp_port = _cfgi('smtp-port')
+    error_from = _cfg('error-from')
+    site_name = _cfg('site-name')
+    if smtp_host and smtp_port and error_from and site_name:
+        mail_handler = SMTPHandler((smtp_host, smtp_port),
+                                   error_from, [error_to],
+                                   site_name + ' Application Exception')
+        mail_handler.setLevel(logging.ERROR)
+        app.logger.addHandler(mail_handler)
+
+# Error handlers. We want to handle errors differently based on their nature, whether it's done via XHR,
+# and whether we're in a debug/development environment.
+
+# app.debug | XHR | HTTPException | What to do?
+# ---------------------------------------------
+# 0         | 0   | 0             | Real 500 -> log, db rollback, error_500.html
+# 0         | 0   | 1             | 4XX -> error_404.html or error_4XX.html
+# 0         | 1   | 0             | Real 500 in API -> log, db rollback, jsonified_exception(e)
+# 0         | 1   | 1             | 4XX in API -> jsonified_exception(e)
+# 1         | 0   | 0             | Real 500 -> log, db rollback, debugger
+# 1         | 0   | 1             | 4XX -> error_404.html or error_4XX.html
+# 1         | 1   | 0             | Real 500 in API -> log, db rollback, jsonified_exception(e)
+# 1         | 1   | 1             | 4XX in API -> jsonified_exception(e)
 
 
 @app.errorhandler(404)
-def handle_404(e: Exception) -> Union[Tuple[str, int], werkzeug.wrappers.Response]:
-    path = request.path
-    if path.startswith('/api/'):
-        return handle_api_exception(e)
-    return render_template("not_found.html"), 404
+def handle_404(e: HTTPException) -> Union[Tuple[str, int], werkzeug.wrappers.Response]:
+    # Switch out the default message
+    if e.description == werkzeug.exceptions.NotFound.description:
+        e.description = "Requested page not found. Looks like this was deleted, or maybe was never here."
+
+    if request.path.startswith("/api") \
+            or (request.accept_mimetypes.accept_json and not request.accept_mimetypes.accept_html):
+        return jsonify_exception(e)
+    return render_template("error_404.html", error=e), 404
+
+
+# This one handles the remaining 4XX errors. JSONified for XHR requests, otherwise the user gets a nice error screen.
+@app.errorhandler(HTTPException)
+def handle_http_exception(e: HTTPException) -> Union[Tuple[str, int], werkzeug.wrappers.Response]:
+    if request.path.startswith("/api") \
+            or (request.accept_mimetypes.accept_json and not request.accept_mimetypes.accept_html):
+        return jsonify_exception(e)
+    return render_template("error_4XX.html", error=e), e.code or 400
```

Comment:
Is this truly limited to only 4XX errors?

https://werkzeug.palletsprojects.com/en/1.0.x/exceptions/

> exception werkzeug.exceptions.InternalServerError(description=None, response=None, original_exception=None)
> 500 Internal Server Error

> ...

>All the exceptions implement this common interface:
> 
> exception werkzeug.exceptions.HTTPException(description=None, response=None)
> Baseclass for all HTTP exceptions. This exception can be called as WSGI application to render a default error page or you can catch the subclasses of it independently and render nicer error messages.

This makes it sound like if we had a reason to call `abort(500)`, it would render `error_4XX.html`.

Reply:
You're right there. We don't call `abort(500)` anywhere right now (or create an `HTTPException` with a code >= 500 or even return >=500 in any form at all).
Not sure if it's a good idea to do that, however we might find a justifying reason at some point, and it's a good idea to be guarded against that.

I think I've found a nice solution to this.
`handle_http_exception()` hands over to `handle_generic_exception()` when the code is >= 500. 

To render the template, `handle_generic_exception()` turns every exception into a `HTTPException` if it isn't already (we still throw away the original message as we did before, for security/privacy reasons).